### PR TITLE
[LIBDOMAIN][FEAT] Reading the schema of objects from the directory service (OpenLDAP)

### PR DESCRIPTION
**Pull request by Ravenwolve**

### CODE REVIEW

File: tests/docker-compose.openldap.local.yml
Comment (august-alt) [inline tests/docker-compose.openldap.local.yml:1]: Remove file from PR.

File: tests/certs/dc0.domain.alt.pem
Comment (august-alt) [inline tests/certs/dc0.domain.alt.pem:1]: Remove file from PR.

File: tests/certs/dc0.domain.alt.key
Comment (august-alt) [inline tests/certs/dc0.domain.alt.key:1]: Remove file from PR.

File: tests/certs/ca.srl
Comment (august-alt) [inline tests/certs/ca.srl:1]: Remove file from PR.

File: tests/certs/ca.pem
Comment (august-alt) [inline tests/certs/ca.pem:1]: Remove file from PR.

File: tests/certs/ca.key
Comment (august-alt) [inline tests/certs/ca.key:1]: Remove file from PR.

**Review by august-alt — CHANGES_REQUESTED** [review]

File: docs/source/INTENDED_USAGE.RU.md
Comment (august-alt) [inline docs/source/INTENDED_USAGE.RU.md:1]: Why do we have this file here? This PR clearly addresses LDAP schema and not documentation.

File: docs/source/INTENDED_USAGE.md
Comment (august-alt) [inline docs/source/INTENDED_USAGE.md:1]: Why do we have this file here? This PR clearly addresses LDAP schema and not documentation.

**Review by august-alt — CHANGES_REQUESTED** [review]

**Review by august-alt — CHANGES_REQUESTED** [review]

File: src/openldap_schema.c [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            if (current_attribute->values == NULL)
+            {
+                continue;
+            }
+
+            int value_index = 0;
+            char* current_value = current_attribute->values[value_index];
+            while (current_value != NULL)
+            {
+                callback(current_value, user_data);
```
[the remark this answers is not in the document]
Comment (Ravenwolve) [reply src/openldap_schema.c:66]: > You need to process the return value of a callback function.

Do you want to return failure if not success?

File: tests/auto/openldap_schema/openldap_schema.c [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    {
+        verto_del(ev);
+
+        char* search_base = "cn=subschema";
+
+        talloc_ctx = talloc_new(NULL);
+
+        schema = ldap_schema_new(talloc_ctx);
+
+        schema_load_openldap(connection, schema);
```
[the remark this answers is not in the document]
Comment (Ravenwolve) [reply tests/auto/openldap_schema/openldap_schema.c:82]: > We have not checked return value of a schema_load_openldap function.

Do you want fail test if not success? And should not we check the schema pointer value?

File: src/connection_state_machine.c [hunk 3]
```diff
@@ -170,6 +171,12 @@ enum OperationReturnCode csm_next_state(struct state_machine_ctx_t *ctx)
             csm_set_state(ctx, LDAP_CONNECTION_STATE_INIT);
         }
         break;
+
+    case LDAP_CONNECTION_STATE_LOAD_SCHEMA:
+        rc = ldap_schema_load(ctx->ctx, ctx->ctx->schema);
+        csm_set_state(ctx, LDAP_CONNECTION_STATE_RUN);
+        break;
+
```
Comment (august-alt) [inline src/connection_state_machine.c:179]: `LDAP_CONNECTION_STATE_LOAD_SCHEMA` should be placed before `LDAP_CONNECTION_STATE_RUN`.

File: src/connection_state_machine.c
Comment (august-alt) [inline src/connection_state_machine.c:1]: Add `LDAP_CONNECTION_STATE_LOAD_SCHEMA` to `const csm_state_value_t state_strings`.

File: src/schema.h [hunk 4]
```diff
@@ -42,4 +43,7 @@ LDAPAttributeType**
 ldap_schema_attribute_types(const ldap_schema_t *schema);
 bool ldap_schema_append_objectclass(ldap_schema_t* schema, LDAPObjectClass *objectclass);
 
+enum OperationReturnCode
+ldap_schema_load(struct ldap_connection_ctx_t* connection, struct ldap_schema_t* schema);
```
Comment (august-alt) [inline src/schema.h:47]: Remove `schema` parameter we will receive schema from connection.

File: src/schema.c [hunk 5]
```diff
@@ -213,5 +222,35 @@ ldap_schema_append_objectclass(struct ldap_schema_t *schema, LDAPObjectClass *ob
...
+        break;
+
+    case LDAP_TYPE_ACTIVE_DIRECTORY:
+        // TODO: move call `schema_load_active_directory` function
+        return RETURN_CODE_SUCCESS;
+        break;
+
+    case LDAP_TYPE_FREE_IPA:
+        // TODO: move call `schema_load_free_ipa` function
+        return RETURN_CODE_SUCCESS;
```
Comment (august-alt) [inline src/schema.c:246]: Replace `RETURN_CODE_SUCCESS` with `RETURN_CODE_FAILURE` for this PR.

File: src/schema.c [hunk 6]
```diff
@@ -213,5 +222,35 @@ ldap_schema_append_objectclass(struct ldap_schema_t *schema, LDAPObjectClass *ob
...
+{
+    switch (connection->directory_type)
+    {
+    case LDAP_TYPE_OPENLDAP:
+        return schema_load_openldap(connection, schema);
+        break;
+
+    case LDAP_TYPE_ACTIVE_DIRECTORY:
+        // TODO: move call `schema_load_active_directory` function
+        return RETURN_CODE_SUCCESS;
```
Comment (august-alt) [inline src/schema.c:241]: Replace `RETURN_CODE_SUCCESS` with `RETURN_CODE_FAILURE` for this PR.

**Review by august-alt — CHANGES_REQUESTED** [review]

File: src/connection_state_machine.c [hunk 7]
```diff
@@ -150,10 +152,15 @@ enum OperationReturnCode csm_next_state(struct state_machine_ctx_t *ctx)
         }
         else
         {
-            csm_set_state(ctx, LDAP_CONNECTION_STATE_RUN);
+            csm_set_state(ctx, LDAP_CONNECTION_STATE_LOAD_SCHEMA);
         }
         break;
 
+    case LDAP_CONNECTION_STATE_LOAD_SCHEMA:
+        rc = ldap_schema_load(ctx->ctx, ctx->ctx->schema);
```
Comment (august-alt) [inline src/connection_state_machine.c:160]: New signature of `ldap_schema_load` does not require a `schema` parameter. We need to remove it.

**Review by august-alt — CHANGES_REQUESTED** [review]

File: src/connection_state_machine.c [hunk 8]
```diff
@@ -150,10 +152,15 @@ enum OperationReturnCode csm_next_state(struct state_machine_ctx_t *ctx)
...
         else
         {
-            csm_set_state(ctx, LDAP_CONNECTION_STATE_RUN);
+            csm_set_state(ctx, LDAP_CONNECTION_STATE_LOAD_SCHEMA);
         }
         break;
 
+    case LDAP_CONNECTION_STATE_LOAD_SCHEMA:
+        rc = ldap_schema_load(ctx->ctx);
+        csm_set_state(ctx, LDAP_CONNECTION_STATE_RUN);
```
Comment (august-alt) [inline src/connection_state_machine.c:161]: Replace
```c
csm_set_state(ctx, LDAP_CONNECTION_STATE_RUN);
```
with
```c            
csm_set_state(ctx, rc == RETURN_CODE_SUCCESS
              ? LDAP_CONNECTION_STATE_RUN
              : LDAP_CONNECTION_STATE_ERROR);
```

**Review by august-alt — CHANGES_REQUESTED** [review]

**Review by august-alt — APPROVED** [review]

**Review by august-alt — APPROVED** [review]

**Review by august-alt — APPROVED** [review]

**Review by august-alt — APPROVED** [review]

**Review by august-alt — APPROVED** [review]

**Pull request merged**